internal/engine/watchmanager.go

```diff
@@ -0,0 +1,154 @@
+package engine
+
+import (
+	"context"
+	"path/filepath"
+
+	"github.com/windmilleng/tilt/internal/ignore"
+	"github.com/windmilleng/tilt/internal/logger"
+	"github.com/windmilleng/tilt/internal/model"
+	"github.com/windmilleng/tilt/internal/store"
+	"github.com/windmilleng/tilt/internal/watch"
+)
+
+type manifestFilesChangedAction struct {
+	manifestName model.ManifestName
+	files        []string
+}
+
+func (manifestFilesChangedAction) Action() {}
+
+type manifestNotifyCancel struct {
+	manifest model.Manifest
+	notify   watch.Notify
+	cancel   func()
+}
+
+type WatchManager struct {
+	watches        map[model.ManifestName]manifestNotifyCancel
+	fsWatcherMaker FsWatcherMaker
+	// TODO(dmiller) do I need a mutex here?
+}
+
+func NewWatchManager(watcherMaker FsWatcherMaker) *WatchManager {
+	return &WatchManager{
+		watches:        make(map[model.ManifestName]manifestNotifyCancel),
+		fsWatcherMaker: watcherMaker,
+	}
+}
+
+func (w *WatchManager) diff(ctx context.Context, state store.EngineState) (setup []model.Manifest, teardown []model.Manifest) {
+	setup = []model.Manifest{}
```

Comment:
can you move the RLockState inside here? you don't really need to read the state after this function completes (we can assume the Manifests in setup and teardown are immutable)

Reply:
Yes sounds good!